WooCommerce/Classes/ServiceLocator/ServiceLocator.swift

```diff
@@ -254,7 +256,8 @@ final class ServiceLocator {
         _cardPresentPaymentsOnboardingIPPUsersRefresher
     }
 
-    static var tapToPayReconnectionController: TapToPayReconnectionController {
+    static var tapToPayReconnectionController: TapToPayReconnectionController<BuiltInReaderConnectionAlertsProvider,
```

Comment:
a question for the future: if we start supporting TapToPay for POS with a different alert provider/presenter, do we need a different singleton of `TapToPayReconnectionController`? I'm guessing it's a singleton for some reason.

Reply:
So this came to mind at the end of yesterday for me as well, and it's a problem we would face in the future with this approach.

Important context – Apple automatically disconnects us from Tap to Pay as soon as the app goes into the background.

It's a singleton because the app starts the reconnection as soon as it comes into the foreground. It only connects if it can do it without any user interaction, but even that takes a few seconds. If the app was backgrounded in the middle of creating an order, you can get to the collect payment step before the connection is finished.

When that happens, we want to adopt the ongoing connection – i.e. start showing alerts for it.

I think to solve it within the existing design, we'd need to make the alert provider and presenter `var` not `let` on the connection controllers, so they can be switched out, and potentially employ some type erasure. It might actually be quite complicated.

A better fix might be to make it possible to adopt ongoing connection attempts lower down the stack, e.g. when we send the `startCardReaderDiscovery` or `connect` actions to the `CardPresentPaymentStore`.

This could be enough of a reason not to take this generic approach now. WDYT?

Reply:
We might also be able to work around the future TTP issue by not adopting the ongoing connection, and instead just waiting for it to finish before any payment is allowed.

Reply:
Thanks for sharing the important context, makes sense to me. As we discussed separately, supporting TTP in POS would be feasible with some refactoring work.